packages/devtools_app/lib/src/screens/performance/panes/raster_stats/raster_stats_model.dart

```diff
@@ -0,0 +1,178 @@
+      layerSnapshots: snapshots,
+      selectedSnapshot: selected,
+      originalFrameSize: originalFrameSize,
+      totalRasterTime: totalRasterTime,
+    );
+  }
+
+  static const _snapshotsJsonKey = 'snapshots';
+
+  static const _selectedIdKey = 'selectedId';
+
+  static const _frameWidthKey = 'frame_width';
+
+  static const _frameHeightKey = 'frame_height';
+
+  final List<LayerSnapshot> layerSnapshots;
+
+  final Size? originalFrameSize;
+
+  final Duration totalRasterTime;
+
+  /// The selected snapshot for this set of raster stats data.
+  ///
+  /// This field is mutable, and is managed by the [RasterStatsController]. It
+  /// is included in [RasterStats] so that it can be encoded in and decoded from
+  /// json.
+  LayerSnapshot? selectedSnapshot;
+
+  Map<String, dynamic> get json => {
+        _frameWidthKey: originalFrameSize?.width.toDouble(),
+        _frameHeightKey: originalFrameSize?.height.toDouble(),
+        _snapshotsJsonKey: layerSnapshots
+            .map((snapshot) => snapshot.json)
+            .toList(growable: false),
+        _selectedIdKey: selectedSnapshot?.id,
+      };
+}
+
+class LayerSnapshot {
+  LayerSnapshot({
+    required this.id,
+    required this.duration,
+    required this.size,
+    required this.offset,
+    required this.bytes,
+  });
+
+  factory LayerSnapshot.parse(Map<String, Object?> json) {
+    final id = json[_layerIdKey] as int;
+    final dur = Duration(microseconds: json[_durationKey] as int);
+    final size = Size(
+      (json[_widthKey] as num).toDouble(),
+      (json[_heightKey] as num).toDouble(),
+    );
+    final offset = Offset(
+      (json[_leftKey] as num).toDouble(),
+      (json[_topKey] as num).toDouble(),
+    );
+    final imageBytes = Uint8List.fromList(
+      (json[_snapshotJsonKey] as List<dynamic>).cast<int>(),
+    );
+    return LayerSnapshot(
+      id: id,
+      duration: dur,
+      size: size,
+      offset: offset,
+      bytes: imageBytes,
+    );
+  }
+
+  static const _layerIdKey = 'layer_unique_id';
+  static const _durationKey = 'duration_micros';
+  static const _snapshotJsonKey = 'snapshot';
+  static const _widthKey = 'width';
+  static const _heightKey = 'height';
+  static const _leftKey = 'left';
+  static const _topKey = 'top';
+
+  final int id;
+
+  final Duration duration;
+
+  final Size size;
+
+  final Offset offset;
+
+  final Uint8List bytes;
+
+  /// The total rendering time for the set of snapshots that this
+  /// [LayerSnapshot] is a part of.
+  ///
+  /// This will be set after this [LayerSnapshot] is created, once all the
+  /// [LayerSnapshot]s in a set have been processed.
+  Duration? totalRenderingDuration;
+
+  double get percentRenderingTimeAsDouble =>
+      duration.inMicroseconds / totalRenderingDuration!.inMicroseconds;
+
+  String get percentRenderingTimeDisplay =>
+      percent2(percentRenderingTimeAsDouble);
+
+  String get displayName => 'Layer $id';
+
+  Map<String, Object?> get json => {
```

Comment:
Instead of manually generating serializers, would we be open to using a serialization package?
https://pub.dev/packages/json_serializable

Reply:
We could certainly look into it. We'll want to make we can use a serialization package in g3 first before making a switch. I'd prefer to do this as part of a larger overhaul, and not as part of this PR, since we are already doing manual serialization in a lot of places in DevTools.

Reply:
Created https://github.com/flutter/devtools/issues/4493 to track.